rts/Game/UI/IKeys.cpp

```diff
@@ -0,0 +1,132 @@
+/* This file is part of the Spring engine (GPL v2 or later), see LICENSE.html */
+
+#include "IKeys.h"
+#include "System/StringUtil.h"
+
+
+int IKeys::GetCode(const std::string& name) const
+{
+	const auto pred = [](const NameCodePair& a, const NameCodePair& b) { return (a.first < b.first); };
+	const auto iter = std::lower_bound(nameToCode.begin(), nameToCode.end(), NameCodePair{name, 0}, pred);
+
+	if (iter == nameToCode.end() || iter->first != name)
+		return -1;
+
+	return iter->second;
+}
+
+
+std::string IKeys::GetName(int code) const
+{
+	const auto pred = [](const CodeNamePair& a, const CodeNamePair& b) { return (a.first < b.first); };
+	const auto iter = std::lower_bound(codeToName.begin(), codeToName.end(), CodeNamePair{0, ""}, pred);
+
+	if (iter == codeToName.end() || iter->first != code)
+		return IntToString(code, "0x%03X");
+
+	return iter->second;
+}
+
+std::string IKeys::GetDefaultName(int code) const
+{
+	const auto pred = [](const CodeNamePair& a, const CodeNamePair& b) { return (a.first < b.first); };
+	const auto iter = std::lower_bound(defaultCodeToName.begin(), defaultCodeToName.end(), CodeNamePair{0, ""}, pred);
+
+	if (iter == defaultCodeToName.end() || iter->first != code)
+		return IntToString(code, "0x%03X");
+
+	return iter->second;
+}
+
+
+bool IKeys::AddKeySymbol(const std::string& name, int code)
+{
+	if ((code < 0) || !IsValidLabel(name))
+		return false;
+
+	const std::string keysym = StringToLower(name);
+
+	// do not allow existing keysyms to be renamed
+	const auto namePred = [](const NameCodePair& a, const NameCodePair& b) { return (a.first < b.first); };
+	const auto nameIter = std::lower_bound(nameToCode.begin(), nameToCode.end(), NameCodePair{keysym, 0}, namePred);
+
+	if (nameIter != nameToCode.end() && nameIter->first == name)
+		return false;
+
+	nameToCode.emplace_back(keysym, code);
+	// assumes that the user would rather see their own names
+	codeToName.emplace_back(code, keysym);
+
+	// swap into position
+	for (size_t i = nameToCode.size() - 1; i > 0; i--) {
+		if (nameToCode[i - 1].first < nameToCode[i].first)
+			break;
+
+		std::swap(nameToCode[i - 1], nameToCode[i]);
+	}
+	for (size_t i = codeToName.size() - 1; i > 0; i--) {
+		if (codeToName[i - 1].first < codeToName[i].first)
+			break;
+
+		std::swap(codeToName[i - 1], codeToName[i]);
+	}
+
+	return true;
+}
+
+
+bool IKeys::IsPrintable(int code) const
+{
+	const auto pred = [](int a, int b) { return (a < b); };
+	const auto iter = std::lower_bound(printableCodes.begin(), printableCodes.end(), code, pred);
+
+	return (iter != printableCodes.end() && *iter == code);
+}
+
+void IKeys::SaveUserKeySymbols(FILE* file) const
+{
+	bool output = false;
+
+	for (const auto& p: nameToCode) {
+		const auto defSymPred = [](const NameCodePair& a, const NameCodePair& b) { return (a.first < b.first); };
```

Comment:
If you use these lambdas across the whole file you might want to just define a couple of comparison functions; either adding them to the classes, or if you don't want to do that you can just add them at the top of this `.cpp`. Then you can reuse them instead of defining lambdas over and over :)

Reply:
Done